language/diem-vm/parallel-executor/src/executor.rs

```diff
@@ -0,0 +1,179 @@
+
+use crate::{
+    errors::*,
+    outcome_array::OutcomeArray,
+    scheduler::Scheduler,
+    task::{ExecutionStatus, ExecutorTask, Transaction, TransactionInferencer},
+};
+use anyhow::Result as AResult;
+use mvhashmap::MVHashMap;
+use num_cpus;
+use rayon::{prelude::*, scope};
+use std::{
+    cmp::{max, min},
+    marker::PhantomData,
+    sync::Arc,
+};
+
+pub struct ParallelTransactionExecutor<T: Transaction, E: ExecutorTask, I: TransactionInferencer> {
+    num_cpus: usize,
+    inferencer: I,
+    phantom: PhantomData<(T, E, I)>,
+}
+
+impl<T, E, I> ParallelTransactionExecutor<T, E, I>
+where
+    T: Transaction,
+    E: ExecutorTask<T = T, Key = T::Key, Value = T::Value>,
+    I: TransactionInferencer<T = T, Key = T::Key>,
+{
+    pub fn new(inferencer: I) -> Self {
+        Self {
+            num_cpus: num_cpus::get(),
+            inferencer,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn execute_transactions_parallel(
+        &self,
+        task_initial_arguments: E::Argument,
+        signature_verified_block: Vec<T>,
+    ) -> Result<Vec<E::Output>, E::Error> {
+        let num_txns = signature_verified_block.len();
+        let chunks_size = max(1, num_txns / self.num_cpus);
+
+        // Get the read and write dependency for each transaction.
+        let infer_result: Vec<_> = {
+            match signature_verified_block
+                .par_iter()
+                .with_min_len(chunks_size)
+                .map(|txn| {
+                    Ok((
+                        self.inferencer.infer_reads(txn)?,
+                        self.inferencer.infer_writes(txn)?,
+                    ))
+                })
+                .collect::<AResult<Vec<_>>>()
+            {
+                Ok(res) => res,
+                Err(_) => return Err(Error::InferencerError),
+            }
+        };
+
+        // Use write analysis result to construct placeholders.
+        let path_version_tuples: Vec<(T::Key, usize)> = infer_result
+            .par_iter()
+            .enumerate()
+            .with_min_len(chunks_size)
+            .fold(Vec::new, |mut acc, (idx, (_, txn_writes))| {
+                acc.extend(txn_writes.clone().into_iter().map(|ap| (ap, idx)));
+                acc
+            })
+            .reduce(Vec::new, |mut lhs, mut rhs| {
+                lhs.append(&mut rhs);
+                lhs
+            });
+
+        let (versioned_data_cache, max_dependency_level) =
+            MVHashMap::new_from_parallel(path_version_tuples);
+        let outcomes = OutcomeArray::new(num_txns);
+
+        let scheduler = Arc::new(Scheduler::new(num_txns));
+
+        scope(|s| {
+            // How many threads to use?
+            let compute_cpus = min(1 + (num_txns / 50), self.num_cpus - 1); // Ensure we have at least 50 tx per thread.
+            let compute_cpus = min(num_txns / max_dependency_level, compute_cpus); // Ensure we do not higher rate of conflict than concurrency.
```

Comment:
Yes, a module to guess/optimize params is probably an overkill (at least for now) but having an util function that takes num_txn, dep level, etc could be good. maybe it could even be worth it to have num_cpus determined there from the library call instead of storing it in the executor (the executor can spawn and run in parallel regardless of how many cpus we have on the machine).

This could be good because it would allow say to later have multiple alternatives for choosing number of (threads) compute_cpus here - e.g. one that uses a constant upper bound (and min logic), one that gets the num_cpus and uses that, one that maybe uses HT (taskset takes care of this easily but still) and/or stays on a socket, etc